### PR TITLE
loading railtoolkit schema via LazyArtifacts

### PR DESCRIPTION
**Pull request by kaat0**
`Path()` and `Train()` validate the data in the constructor with JsonSchema.
Introduced two new functions `load()` and `get_schema()`, to change from hard-coded JsonSchema to using the JsonSchema via LazyArtifacts.

`Path()` and `Train()` now use `load()` for a file.  `load()` returns a validated Dict() according to the files specified JsonSchema (which has to be registered beforehand in the Artifacts.toml file).

Dependencies will also be needed for issue #11.

### CODE REVIEW

File: Artifacts.toml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: UTF-8 -*-
+# __author__        = "Martin Scheidt"
+# __copyright__     = "2024"
+# __license__       = "ISC"
+
+## Julia commands to get hashes:
+# using Tar, Inflate, SHA
+# filename = "railtoolkit-2022.05.tar.gz"
+# println("sha256: ", bytes2hex(open(sha256, filename)))
+# println("git-tree-sha1: ", Tar.tree_hash(IOBuffer(inflate_gzip(filename))))
```
Comment (gwehrle) [inline Artifacts.toml:10]: Its pretty neat to have a comment how to add more artifacts.
But I would only reference using ArtifactsUtils.jl or what do you think?

File: src/constructors.jl [hunk 2]
```diff
@@ -4,6 +4,117 @@
...
+"""
+function get_schema(data::Dict)::Schema
+    ## assumption: schema come in the form of (http://)schema_name/schema_subtype
+
+    provided_schema = data["schema"]
+    @debug "loading file - detected schema: $provided_schema"
+    schema_version = lowercase(data["schema_version"])
+    @debug "loading file - detected schema version: $schema_version"
+
+    schema = lowercase(provided_schema)
```
Comment (gwehrle) [inline src/constructors.jl:88]: Parsing URLs manually can be really tricky. I would recommend using https://github.com/JuliaWeb/URIParser.jl to do this!

File: src/constructors.jl [hunk 3]
```diff
@@ -123,7 +234,7 @@ julia> my_path = Path("file.yaml")
 Path(variables)
 ```
 """
-function Path(file, type = :YAML)
+function Path(file::String)
 
     ## default values
     name = ""
```
Comment (gwehrle) [inline src/constructors.jl:240]: are these defaults still needed?

**Review by gwehrle — COMMENTED** [review]
Pretty nice!
I had some small adjustments. Please also run JuliaFormatter ;) Also some method descriptions are out of date.

File: test/file_loading.jl [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    logger = ConsoleLogger(stderr, Warn)
+    with_logger(logger) do
+        @test_throws ErrorException("Can not load 'data/paths/missing_schema.yml'. No attribute with 'schema' found!") TrainRuns.load("data/paths/missing_schema.yml")
+        @test_throws ErrorException("Can not load 'data/paths/missing_schema_version.yml'. No attribute with 'schema_version' found!") TrainRuns.load("data/paths/missing_schema_version.yml")
+        @test_throws ErrorException("Can not load 'data/paths/broken.xml'. Unsupported file extension!") TrainRuns.load("data/paths/broken.xml")
+        @test_throws ErrorException("Can not load 'data/paths/broken.yaml'. Format not recognized!") TrainRuns.load("data/paths/broken.yaml")
+        @test_throws ErrorException("Can not load 'data/paths/missing_path.yml'. Format not recognized!") TrainRuns.load("data/paths/missing_path.yml")
+    end
+end
+#
```
[the remark this answers is not in the document]
Comment (kaat0) [reply test/file_loading.jl]: No, just a typo. I removed it.

File: Artifacts.toml (hunk 1 above)
Comment (kaat0) [reply Artifacts.toml:10]: Agree!

File: src/constructors.jl (hunk 3 above)
Comment (kaat0) [reply src/constructors.jl:240]: What do you mean?

Comment (gwehrle) [reply src/constructors.jl:240]: Variables are like `name` and `id` are defined as empty string just to overwrite them a few lines later. I think the first definition with empty string is not necessary.

File: src/constructors.jl (hunk 2 above)
Comment (kaat0) [reply src/constructors.jl]: Resolved in [4aa0fef](https://github.com/railtoolkit/TrainRuns.jl/pull/81/commits/4aa0fefc8e7ce9ca109d68e03003b24ddad6f903)

File: src/constructors.jl (hunk 3 above)
Comment (kaat0) [reply src/constructors.jl]: resolved with [eb6c37e](https://github.com/railtoolkit/TrainRuns.jl/pull/81/commits/eb6c37e8f3459d1ff504427b2726fd423fcec8d7)

File: test/artifact_schema_loading.jl
[the remark this answers is not in the document]
Comment (kaat0) [reply test/artifact_schema_loading.jl:1]: resolved in [3678412](https://github.com/railtoolkit/TrainRuns.jl/pull/81/commits/36784126605554f660876f9ddbb06c356dd99c8f)

**Comment by kaat0** [comment]
Currently two targets fail:
- 32bit architecture (x86)
- long term support (lts) julia with 1.6

**Pull request merged**